contrib/golang/upstreams/http/tcp/source/upstream_request.cc

```diff
@@ -0,0 +1,479 @@
+        fmt::format("golang filter failed to parse plugin config: {} {}", so_id_, so_path_));
+  }
+
+  ENVOY_LOG(debug, "tcp upstream new plugin config, id: {}", config_id_);
+}
+
+TcpConnPool::TcpConnPool(Upstream::ThreadLocalCluster& thread_local_cluster, 
+  Upstream::ResourcePriority priority, Upstream::LoadBalancerContext* ctx, const Protobuf::Message& config) {
+    conn_pool_data_ = thread_local_cluster.tcpConnPool(priority, ctx);
+
+    envoy::extensions::upstreams::http::tcp::golang::v3alpha::Config c;
+    auto s = c.ParseFromString(config.SerializeAsString());
+    ASSERT(s, "any.value() ParseFromString should always successful");
+
+    ENVOY_LOG(debug, "tcp upstream load tcp_upstream_golang library at parse config: {} {}", c.library_id(), c.library_path());
+
+    dynamic_lib_ = Dso::DsoManager<Dso::TcpUpstreamDsoImpl>::load(
+      c.library_id(), c.library_path(), c.plugin_name());
+    if (dynamic_lib_ == nullptr) {
+      throw EnvoyException(fmt::format("tcp upstream : load library failed: {} {}", c.library_id(), c.library_path()));
+    };
+
+    FilterConfigSharedPtr conf = std::make_shared<FilterConfig>(c, dynamic_lib_);
+    conf->newGoPluginConfig();
+    config_ = conf;
+    plugin_name_ = c.plugin_name();
+}
+
+void TcpConnPool::onPoolReady(Envoy::Tcp::ConnectionPool::ConnectionDataPtr&& conn_data,
+                              Upstream::HostDescriptionConstSharedPtr host) {
+  upstream_handle_ = nullptr;
+  Network::Connection& latched_conn = conn_data->connection();
+  auto upstream =
+      std::make_unique<TcpUpstream>(&callbacks_->upstreamToDownstream(), std::move(conn_data), dynamic_lib_, config_);
+
+  ENVOY_LOG(debug, "tcp upstream get host info: {}", host->cluster().name());
+
+  callbacks_->onPoolReady(std::move(upstream), host, latched_conn.connectionInfoProvider(),latched_conn.streamInfo(), {});       
+}
+
+TcpUpstream::TcpUpstream(Router::UpstreamToDownstream* upstream_request,
+                         Envoy::Tcp::ConnectionPool::ConnectionDataPtr&& upstream, Dso::TcpUpstreamDsoPtr dynamic_lib,
+                         FilterConfigSharedPtr config)
+    :encoding_state_(new EncodingProcessorState(*this)), decoding_state_(new DecodingProcessorState(*this)),
+     route_entry_(upstream_request->route().routeEntry()), upstream_request_(upstream_request), upstream_conn_data_(std::move(upstream)), dynamic_lib_(dynamic_lib), config_(config) {
+  
+  configId = config->getConfigId();
+  plugin_name.data = config->pluginName().data();
+  plugin_name.len = config->pluginName().length();
+
+  upstream_conn_data_->connection().enableHalfClose(true);
+  upstream_conn_data_->addUpstreamCallbacks(*this);
+}
+
+TcpUpstream::~TcpUpstream() {
+  ENVOY_LOG(debug, "tcp upstream on destroy");
+
+  // initRequest haven't be called yet, which mean haven't called into Go.
+  if (configId == 0) {
+    return;
+  }              
+
+  auto reason = (decoding_state_->isProcessingInGo() || encoding_state_->isProcessingInGo())
```

Comment:
it won't be processingInGo, we'd better add an `assert` here instead.

Reply:
done